feat(blob): update approach to remove reliance on column groups, break down plan (#18013)
apache/hudi: rfc/rfc-100/rfc-100.md

### Building on Existing Foundation

This RFC leverages foundation pieces:

1**RFC-99 BLOB Types**: Introduces BLOB types to the Hudi type system, providing the type foundation for unstructured data storage.

## Requirements 

Below are the high-level requirements for this feature.

1. Users must be able to define tables with a mix of structured (current types) and unstructured (blob type)
   columns
2. Unstructured data can be stored inline (e.g small images right inside the file) or out-of-line (e.g

>    pointer to a multi-GB video file someplace). Out-of-line references can also include a position within the file which 

Reviewer: ```suggestion
   pointer to a multi-GB video file someplace). Out-of-line references can also include a (position, length) within the file which 
```

   allows multiple blobs to be stored within a single file to reduce the number of files in storage. 
   This decision should be made dynamically during write/storage time.
3. All table life-cycle operations and table services work seamlessly across both column types.for e.g cleaning
   the file slices should reclaim out-of-line blob data when the reference is managed by Hudi. 
   Clustering should be able re-organize records across file groups or even repack blobs if required.
4. Hudi should expose controls at the writer level, to control whether to store blobs inline or out-of-line
   based on size thresholds.
5. Query engines like Spark should be able to read the unstructured data and materialize the values lazily to reduce memory pressure during shuffles.

## High-Level Design

The design introduces an abstraction that allows inline and out-of-line storage of byte arrays that work seamlessly for the end user. Structured columns continue using 
Parquet format, while unstructured data can use specialized formats like Lance or optimized Parquet configurations or HFile for random-access.

### 1. Storage Abstraction

> We will add a Blob type to the HoodieSchema that encapsulates both inline and out-of-line storage strategies. This will allow the user to use a mix of storage strategies seamlessly.

Reviewer: ```suggestion
We will add a `blob` type to the HoodieSchema that encapsulates both inline and out-of-line storage strategies. This will allow the user to use a mix of storage strategies seamlessly.
```

Reviewer: this will be different than the current `binary` type, which will always be inline. correct?

Author: Yes, it will be different. I am wondering if we want to include a schema evolution option to evolve binary fields to blobs when users upgrade their tables.

**Storage Schema**:
```json
{
  "type": "record",
  "name": "Blob",
  "fields": [
    {"name": "storage_type", "type": "string"},
    {"name": "data", "type": ["null", "bytes"]},
    {"name": "reference", "type": ["null", {
      "type": "record",
      "name": "BlobReference",
      "fields": [

>         {"name": "external_path", "type": "string"},

Author: I was imagining always fully qualified. Would `blobBasePath` be a sub-directory of the table path? Are you imagining it as the storage location for the files written by Hudi?

>         {"name": "external_path", "type": "string"},

Reviewer: fully qualified is okay for now. `blobBasePath` can be anywhere, not necessarily sub folder.

>         {"name": "position", "type": "long"},

Reviewer: use `offset` and `length` ? (just naming comment)

        {"name": "size", "type": "long"},
        {"name": "managed", "type": "boolean"}
      ]
    }]}
  ]
}
```
The `managed` flag will be used by the cleaner to determine if an out-of-line blob should be deleted when cleaning up old file slices. This allows users to point to existing files without Hudi deleting them.

### 2. Reader
Readers will be updated to allow for lazy loading of the blob data, even when it is inline. This will help reduce memory pressure during shuffles in distributed query engines like Spark.

> The readers will return a reference to the blob data in the form of a path, position, and size. This applies for both inline and out-of-line storage.

Author: Yes, that is my hope but some initial investigation with Lance showed this is not possible yet with the java library.

> The reader user will then use a UserDefinedFunction (UDF), or similar abstraction based on the engine, to read the blob data from the reference when needed.

Reviewer: Might be worth the effort to sketch what an efficient `.materialize()` UDF on a blob column looks like.. ensure, there are sufficient APIs in Spark to cover that.

### 3. Writer
#### Phase 1: Basic Blob Support
The writer will be updated to support writing blob data in both inline and out-of-line formats. 
For out-of-line storage, the assumption is that the user will provide the external path, position, and size of the blob data and these references will not be managed by Hudi (they are not removed by the cleaner).
In this phase, we will not implement dynamic inline/out-of-line storage based on size thresholds.

> The writer should be flexible enough to allow the user to pass in a dataset with blob data as simple byte arrays or records matching the Blob schema defined above.

Reviewer: ok

#### Phase 2: Dynamic Inline/Out-of-Line Storage

> In this phase, the writer will be updated to support dynamic inline/out-of-line storage based on user configured size thresholds. The user will still be able to provide their own external path for out-of-line storage if desired.

Reviewer: > the user will still be able to provide their own external path for out-of-line storage if desired

IIUC here, the blob will just be a reference.. Only the blob data in `blob_col.data` will be subject to these size thresholds.

When the user provides blob data in the form of byte arrays, the writer will take arrays larger than the configured threshold and write them to files. The user can configure the file type used for this storage (e.g Parquet, HFile, Lance, etc).
The writer will then generate the appropriate BlobReference for the out-of-line storage and write that to the main file.
Multiple blobs can be written to the same file to reduce the number of files created in storage. All of these blobs will belong to the same file group for ease of management.

**Configurations**: 
- `hoodie.storage.blog.inline.threshold`: Size threshold in bytes for inline vs out-of-line storage

> - `hoodie.storage.blob.outofline.packing.threshold`: Size threshold in bytes for blobs that can be packed together in a single out-of-line file

Author: I like container, will update naming and wording here

- `hoodie.storage.blob.outofline.packing.maxFileSize`: Size threshold in bytes for maximum size of an out-of-line blob file
- `hoodie.storage.blob.outofline.format`: File format to use for out-of-line blob storage

**External Storage Layout**:
```

> {table_path}/.hoodie/blobs/{partition}/{file_group_id}/{column_name}/{instant}/{blob_id}

Reviewer: see comment above on whether we should make this location configurable via a table prop.

Reviewer: Another comment is - this feels a little deep in terms of folder hierarachy. but if we will simply do `GET` requests against it, may be its okay. 

Given we have the file reference, can we just leave this at the `partition` level and then may be some additional sub folders if we reach some threshold. say 10K blobs under `{table_path}/.hoodie/blobs/{partition}` . 

I don't have an informed suggestion yet, but spirit is to avoid anything that will get us into a deeply nested hierarchy trouble, and be able to handle millions of blobs stored under a prefix.. For e.g we have explored in RFC-60, how to add a random prefix to force sth like S3 to distribute these blobs more evenly..

```

#### Writer Optimizations for Blob Storage
##### Parquet

> For unstructured column groups using Parquet:

Author: The presentation is a bit light on details, the main takeaways not already included here are smaller pages and bigger row groups if you want them to help with random access

> For unstructured column groups using Parquet:

Reviewer: can be done later. will keep comment open for later

- **Disable Compression**: Avoid double compression of already compressed media files
- **Plain Encoding**: Use PLAIN encoding instead of dictionary encoding for BLOB columns
- **Large Page Sizes**: Configure larger page sizes to optimize for sequential BLOB access
- **Metadata Index**: Maintain BLOB metadata in Hudi metadata table for efficient retrieval of a single blob value.
- **Disable stats**: Not very useful for BLOB columns

> ##### Lance

Reviewer: Can we file the gaps you found as issues.. (if not done already)

- Native support for high-dimensional vectors and embeddings
- Efficient columnar storage for mixed structured/unstructured data
- Better compression for certain unstructured data types

Supporting Lance, working across Hudi + Lance communities will help users unlock benefits of both currently supported
file formats in Hudi (parquet, orc), along with benefits of Lance. Over time, we could also incorporate newer emerging
file formats in the space and other well-established unstructured file formats.

### 4. Table Services
#### Cleaning
The cleaning service will be updated to identify the out-of-line blob references that are managed by Hudi and no longer referenced by any active file slices.
To identify these references, we have two options:
1. Scan all active file slices to build a set of referenced blob IDs and then scan the file slices being removed to identify references in the removed slices that are not in the active set.

> 2. Maintain metadata on the blob references contained in the file in the footer or metadata section of each base and log file. The cleaner can then read this metadata to identify blob references in the removed slices and check if they are still referenced in active slices.

Author: True, to start I think it will be best to do the brute force implementation (option 1) and then design the optimization. 

For this option, I was imagining the file only referencing a handful of container files.

3. Maintain an index in the metadata table that tracks all blob references and their reference counts. The cleaner can then use this index to identify unreferenced blobs.

**Note**: This is only required for out-of-line blobs that are managed by Hudi. Out-of-line blobs that are not managed by Hudi will not be deleted by the cleaner. This will be part of `Phase 2` of the writer implementation.

#### Blob Compaction
We will introduce a new form of compaction that allows for repacking of out-of-line blobs managed by Hudi to reduce the number of files in storage.

> The compaction will scan the out-of-line blob references that are active within the file group and repack them into new files based on a user configured target file size.

Reviewer: ```suggestion
The compaction will scan the out-of-line blob references that are active within the file group and repack them into new packed files based on a user configured target file size.
```

Reviewer: Similar to current compaction, blob compaction does not alter the record key to fg mapping. correct?

The repacking will also pack these blobs based on the base file row's ordering to improve read locality.

#### Clustering
Since the out-of-line blobs are part of the file group, managed references will need to be updated as part of the clustering operation.

> The clustering will need to read the blob references from the source files and rewrite them to new files. 

Reviewer: ```suggestion
The clustering will need to read the blob references from the source file groups and rewrite them to new target file groups. 
```

- Efficient BLOB streaming for distributed ML workloads
- Integration with Ray's object store for large BLOB caching

### 6. Metadata Table Extensions

- Track BLOB references for garbage collection
- Store maintain indexes for parquet based blob storage
- Maintain size statistics for storage optimization
- Support BLOB-based query optimization

## Development Plan

> #### Milestone 1: External Blob Support

Author: I don't think UDF is the right approach here since that will act on a single row and remove the possibility of batching the lookups when we have these blob container files. See the reader section above for the overview and this [PR](https://github.com/apache/hudi/pull/18098) for the implementation.

I'm updating the reader section to add the expectations on the reference returned

- Efficient BLOB streaming for distributed ML workloads
- Integration with Ray's object store for large BLOB caching

### 6. Metadata Table Extensions

> - Track BLOB references for garbage collection

Reviewer: remove some of these? not relevant anymore
